pkg/antctl/raw/upgrade/apistorage/command.go

```
@@ -0,0 +1,260 @@
+		}
+		// Only upgrade the CRDs with suffix "crd.antrea.io".
+		crdNamesToUpgrade = crdNamesToUpgrade.Intersection(antreaCRDNames)
+	}
+	return crdNamesToUpgrade, nil
+}
+
+// getCRDsToUpgrade gets a list of Antrea CRDs to upgrade.
+func getCRDsToUpgrade(cmd *cobra.Command, k8sClient client.Client, crdNameListToUpgrade sets.Set[string]) []*apiextv1.CustomResourceDefinition {
+	var crdsToUpgrade []*apiextv1.CustomResourceDefinition
+	for crdName := range crdNameListToUpgrade {
+		crd := &apiextv1.CustomResourceDefinition{}
+		if err := k8sClient.Get(context.TODO(), client.ObjectKey{Name: crdName}, crd); err != nil {
+			fmt.Fprintf(cmd.ErrOrStderr(), "Failed to get CRD %q: %v.\n", crdName, err)
+			continue
+		}
+		// Skip the CRD that has only one version.
+		if len(crd.Spec.Versions) < 2 {
+			fmt.Fprintf(cmd.OutOrStdout(), "Skip upgrading CRD %q since it only has one version.\n", crdName)
+			continue
+		}
+		crdsToUpgrade = append(crdsToUpgrade, crd)
+	}
+	return crdsToUpgrade
+}
+
+// upgradeCRDObjects upgrades the existing objects of a CRD.
+func upgradeCRDObjects(cmd *cobra.Command, k8sClient client.Client, crd *apiextv1.CustomResourceDefinition) error {
+	list := &unstructured.UnstructuredList{}
+	list.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   crd.Spec.Group,
+		Version: getCRDStorageVersion(crd),
+		Kind:    crd.Spec.Names.ListKind,
+	})
+	if err := k8sClient.List(context.TODO(), list); err != nil {
+		return err
+	}
+	if len(list.Items) == 0 {
+		return nil
+	}
+	fmt.Fprintf(cmd.OutOrStdout(), "Upgrading %d objects of CRD %q.\n", len(list.Items), crd.Name)
+	for _, obj := range list.Items {
+		copyObj := obj
+		err := retry.OnError(wait.Backoff{
+			Duration: time.Second,
+			Steps:    5,
+		}, func(err error) bool {
+			return validateErr(err) != nil
+		}, func() error {
+			return k8sClient.Update(context.TODO(), &copyObj)
+		})
+		if validateErr(err) != nil {
+			return err
+		}
```

I don't think it should ignore any error but NotFound here, as it means upgrading this object didn't succeed, but the following handings all treat them as success. Even the return code would be 0. This is undesired.

---

Add a counter to count failed objects.